BUG: Fix eigh and cholesky methods of numpy.random.multivariate_normal

Balandat opened the pull request:
Fixes #15871. The `eigh` and `cholesky` methods were not producing results with the correct covariance.

Balandat committed:
Bug: Fix eigh mnd cholesky methods of numpy.random.multivariate_normal

Fixes #15871

eric-wieser commented:
Needs a test. Would be worth checking why CI is failing too.

charris commented on numpy/random/_generator.pyx:
```diff
@@ -3535,7 +3535,7 @@ cdef class Generator:
             # if check_valid == 'ignore' we need to ensure that np.sqrt does not
             # return a NaN if s is a very small negative number that is
             # approximately zero or when the covariance is not positive-semidefinite
-            _factor = u * np.sqrt(abs(s))
```
Looking at this more carefully, `cov = (u * sqrt(s)) @ (u * sqrt(s)).T` so I think this is correct.

Balandat replied:
Yes this is correct for the covariance (`_factor @ _factor.T = cov`), but then the `np.dot(x, _factor)` below is incorrect. In particular, for the `svd` version, you will realize that `_factor @ factor.T != cov`, but instead  `_factor.T @ factor = cov`.

Assuming that `_factor @ factor.T = cov`, what you want is 
```
E[(_factor @ x.T) @ (_factor @ x.T).T]
= E[_factor @ (x.T @ x) @ _factor.T]
= _factor @ E[x.T @ x] @ factor.T
= _factor @ I @ _factor.T
= cov
```
where for the single sample `x` here is `1 x n` with `n` the size of the covariance matrix. 
So the correlated samples here should be `_factor @ x.T` and not `np.dot(x, _factor) = x @ _factor`, otherwise what you getting instead now is sth like
```
E[(x @ _factor) @ (x @ _factor).T]
= E[x @ (_factor @ _factor.T) @ x.T]
= E[x @ cov @ x.T]
```

So two possible changes: 
1. Change the `_factor` definition for `eigh` as done in this one line diff and keep using `x @ _factor` as the correlated samples
2. Change the `_factor` definition for the `svd` and `cholesky` methods to be consistent s.t. `_factor @ factor.T = cov`, and use `_factor @ x` for the correlated samples throughout.

charris replied:
Yeah, I was just thinking I should check the rest of the implementation. The first way is the least intrusive, OTOH, the second is how I think about the problem and may be more easily extensible.  For now, your PR is the simplest fix.

charris commented:
@eric-wieser 

Because the generators may change, the simplest test might be some rough statistics on the result. What do you suggest?

Balandat commented:
Re the test: We can just use the singular matrix from my example and then check that the first two samples are approx. equal for `svd` and `eigh`. Not sure what other things are tested already, but we should probably at least do some checks on the (marginal) variances and means if that is not done yet.

I do think that we should be consistent about the factor - typically the correlation is done as `y = F @ x` s.t. `E[y y.T] = E[F @ x @ x.T @ F.T] = F @ E[x @ x.T] @ F.T = F @ F.T`, so we should have `F @ F.T = cov`. 

Note also that if many evals are zero, we could remove these evecs (or columns of U,V for svd) and work with sth like the following:
```
def eigh_sample(mean, cov, cutoff=1e-10):
    evals, evecs = np.linalg.eigh(cov)
    pos = evals > cutoff
    evecs_pos = evecs[:, pos]
    base_samples = rg.normal((np.sum(pos), 1))
    scale = np.sqrt(evals[pos])
    return mean + (scale * evecs_pos) @ base_samples
```
Not sure if it's worth doing this if we re-compute the eigendecomposition/svd every time though.

charris commented:
> Not sure if it's worth doing this if we re-compute the eigendecomposition/svd every time though.

There was some discussion of adding the option of passing the left-hand factor, but that actually seems more appropriate to a class based implementation, that is, one could ask for a generator that returned multivariate normals with given covariance. If I needed a bunch of variates, that is how I would do it, but that might not be appropriate for NumPy.

Balandat commented:
I agree this probably doesn't belong into numpy. I'm surprised scipy's mvn for "frozen" rvs doesn't deal with this, looks like the matrix decomposition is recomputed for each sample here: https://github.com/scipy/scipy/blob/v1.4.1/scipy/stats/_multivariate.py#L660

Torch's MVN does store the factor (though it doesn't allow singular matrices): https://github.com/pytorch/pytorch/blob/master/torch/distributions/multivariate_normal.py#L146-L15

charris commented:
> I'm surprised scipy's mvn for "frozen" rvs doesn't deal with this

Might be worth opening a SciPy issue.

charris commented:
> but we should probably at least do some checks on the (marginal) variances and means if that is not done yet.

That is what I was thinking. It shouldn't require too large a sample to get sufficient accuracy to catch big errors like the current one. I would suggest starting with an `L` and passing `L @ L.T`

Balandat commented:
> Might be worth opening a SciPy issue.

https://github.com/scipy/scipy/issues/11772

seberg commented on numpy/random/_generator.pyx:
```diff
@@ -3535,7 +3535,7 @@ cdef class Generator:
             # return a NaN if s is a very small negative number that is
             # approximately zero or when the covariance is not positive-semidefinite
-            _factor = u * np.sqrt(abs(s))
+            _factor = np.transpose(u * np.sqrt(abs(s)))
```
I am a bit worried that we silently accept more than 2 dimensions here (what happens for more than 2 dimensions? If it is an error, just having a test for that could be fine as well, maybe with a comment that the code may need updating), but than transpose would have to transpose only the last two dimensions, maybe `np.swapaxes` is actually good (for once).

The main point is, we really need at least some test, a random/fuzzy one that is lenient enough to not fail would seem fine to me. Or maybe one that checks approximate consistency over the different `methods` at least?

Balandat replied:
Agree, we can be explicit about transposing the last two dimensions. Also agree on the tests, will add some this weekend.

charris replied:
> but than transpose would have to transpose only the last two dimensions

That is what I meant about the standard form being easier to extend. But I think we can deal with that larger refactoring if we ever decide to go the stacked route.

eric-wieser replied:
> but than transpose would have to transpose only the last two dimensions

The pseido-private `linalg.transpose` does this.

Balandat commented:
Changed to transpose on trailing dims. Added generator-agnostic test for singular covariances. Maybe someone else can take on additional tests checking some basic statistics on the generated samples in a separate PR.

charris commented:
I think the test is too special cased. I think that a rough check the statistics all methods would be useful. For instance, looks like cholesky also yields wrong results.
```
In [36]: b = default_rng(123).multivariate_normal([0, 0], cov, size=(1000,), method='svd')                                              

In [37]: (b.T @ b)/1000                                                                                                                 
Out[37]: 
array([[1.06677611, 1.05240103],
       [1.05240103, 2.04555217]])

In [38]: b = default_rng(123).multivariate_normal([0, 0], cov, size=(1000,), method='eigh')                                             

In [39]: (b.T @ b)/1000                                                                                                                 
Out[39]: 
array([[ 0.38497443, -0.0150749 ],
       [-0.0150749 ,  2.79194006]])

In [40]: b = default_rng(123).multivariate_normal([0, 0], cov, size=(1000,), method='cholesky')                                         

In [41]: (b.T @ b)/1000                                                                                                                 
Out[41]: 
array([[2.14015452, 1.06968605],
       [1.06968605, 1.03675997]])

In [42]: cov                                                                                                                            
Out[42]: 
array([[1., 1.],
       [1., 2.]])

In [43]: numpy.__version__                                                                                                              
Out[43]: '1.19.0.dev0+f1a247f'
```

Balandat commented:
Oh so the Cholesky samples are broken too. We may as well just go all in on the `_factor @ x` representation then and only transpose the svd factor

charris commented:
We will need a release note to go with this. You can see how to do this by looking at the current release note files in `doc/release/upcoming_changes/`.

Balandat commented:
Changed representation internally. Was easiest to use `einsum` for the output, hopefully that's ok. Added a test that is pretty loose and should only catch significant errors. At a later stage one could do some proper hypothesis testing with an extremely low significance level to avoid flaky tests.
Added upcoming_changes.

mattip commented:
Should the tests cover the new `numpy.random.default_rng` as well, or are there already similar tests elsewhere?

Balandat commented:
not sure about other rngs, maybe someone who knows the codebase better can clarify?

charris commented on numpy/random/_generator.pyx:
```diff
@@ -3537,10 +3537,9 @@ cdef class Generator:
             _factor = u * np.sqrt(abs(s))
         else:
-            _factor = np.sqrt(s)[:, None] * vh
+            _factor = u * np.sqrt(s)
```
The problem here is that `u` is not necessarily `vh.T` for the `svd`, if there are negative eigenvalues from numerical error, there will be a sign change. Hmm... if the cov is valid and the negative eigenvalue is due to numerical error, i.e., it is really tiny, then it should have hardly any impact on the output, so this is probably OK. Might mention the possibility in the release note.

Balandat replied:
I mean the same is true for the previous formulation, right?

charris replied:
Yep, it would give incorrect results in that case, but different incorrect results :)

Balandat replied:
ok let's make sure our errors are correct then

Balandat commented on numpy/random/tests/test_generator_mt19937.py:
```diff
@@ -1242,6 +1242,28 @@ def test_multivariate_normal(self, method):
         assert_raises(ValueError, random.multivariate_normal, mean, cov,
                       check_valid='raise', method='eigh')
 
+        # check degenerate samples from singular covariance matrix
```
I mean using a fixed seed and checking the samples will do the trick for validating correctness of the formulation. The statistics test is really more and end-to-end integration test. That should really be done in a "statistical unit testing" setting where you run the test a bunch of times under a hypothesis test and then see that you get reasonably close to your selected significance level across many runs. But that shouldn't be part of the regular test suite.

charris replied:
> But that shouldn't be part of the regular test suite.

I would support adding a validation test suite at some point. But for now we could keep this and move it later. It did uncover the Cholesky problem and would have caught the `eigh` problem.

The various random functions were tested for statistical accuracy many years ago, we should really have tests somewhere that we could use to revalidate the results since there have been changes since. Hmm, that could almost be a GSOC project next time that rolls around, not that statistical tests are the least bit trivial but the problem is well stated. Validating the generators themselves might be worth something just to make sure we didn't drop the ball somewhere, after all, there was the buggy generation of low precision integers that we should have caught earlier.

@bashtage @rkern Thoughts?

rkern replied:
Agreed.

charris commented on numpy/random/_generator.pyx:
```diff
@@ -3537,10 +3537,9 @@ cdef class Generator:
-        x = np.dot(x, _factor)
-        x += mean
+        x = mean + np.einsum("ij,...j->...i", _factor, x)
         x.shape = tuple(final_shape)
```
BLAS will be faster than einsum. `x` will always be 2D at this point due to previous reshape and transpose is cheap, so just `x @ _factor.T` should be fine and at least keeps all the transposes in one place while keeping the results the same for the 'svd' case. The output shape is what make things tricky. On could transpose `x` in the reshape and do `(_factor @ x).T` but that doesn't seem much of an improvement.

Balandat replied:
Ah I see I thought `x` here could still have a bunch of leading dims. That makes things easier.

Balandat commented:
Using `x @ _factor.T`, splitting out the basic stats test & relaxing tolerances.

charris commented on numpy/random/tests/test_generator_mt19937.py:
```diff
@@ -1259,6 +1270,18 @@ def test_multivariate_normal(self, method):
+    def test_multivariate_normal_basic_stats(self, method):
+        n_s = 1000
+        mean = np.array([1, 2])
+        cov = np.array([[2, 1], [1, 2]])
```
```
random = Generator(MT19937(self.seed))
```
Otherwise it uses the global unseeded generator defined at the top of the file. We should probably get rid of that at some point.

charris commented on numpy/random/tests/test_generator_mt19937.py:
```diff
@@ -1259,6 +1270,18 @@ def test_multivariate_normal(self, method):
+        cov_emp = (s_center.T @ s_center) / (n_s - 1)
+        # these are pretty loose and are only designed to detect major errors
+        assert np.all(np.abs(s_center.mean(-2)) < 0.2)
+        assert np.all(np.abs(cov_emp - cov) < 0.4)
```
This could be as tight as before if it works with the seeded generator it should be repeatable.

charris commented:
Looks good, let's see if the tests pass.

charris commented:
Thanks @Balandat. It is very nice to have those errors fixed and tested.

bashtage commented:
@charris It would be a good idea. @WarrenWeckesser discovered a casting issue using some Monte Carlo methods IIRC.  It would make sense to start with the things that are "newer", e.g., the Ziggurat exponentials, gamma, normals, and the bounded integers.  These are all widely used in other distributions.
